### DISCUSSION TITLE
Duplicate Key error, need Update instead of Insert

### DISCUSSION THREAD

**Discussion by santalvarez** [sqlalchemy/sqlalchemy#9252] [Usage Questions] [answered]
Hi, I have these three objects **Device**, **Host** and **OperatingSystem**.

```python
class Device(Base):
    __tablename__ = "device"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    serial_number = Column(String(100), nullable=False, unique=True)
    first_seen = Column(DateTime, nullable=False, default=datetime.utcnow)

    # One-to-one relationship with Host
    host = relationship("Host", back_populates="device", uselist=False, cascade="all, delete-orphan")

class Host(Base, Component):
    __tablename__ = "host"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    device_id = Column(BigInteger, ForeignKey("device.id"), unique=True, nullable=False)
    os_id = Column(BigInteger, ForeignKey("operating_system.id"))
    hostname = Column(String(20), nullable=True)

    # Many Hosts to one OS
    os = relationship("OperatingSystem", back_populates="hosts")
    device = relationship("Device", back_populates="host", uselist=False)

class OperatingSystem(Base):
    __tablename__ = "operating_system"
    __table_args__ = (UniqueConstraint("name", "version", "patch", name="os_uc"),)

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    name = Column(Enum(OperatingSystemName), nullable=False)
    version = Column(String(30), nullable=False)
    patch = Column(String(30), default="", nullable=False)

    # One OS to many Hosts
    hosts = relationship("Host", back_populates="os")

```

I am trying to add a **Host** to an existing **Device**. Previous to adding the **Host** I am checking the **OS** exists on the db to avoid a _duplicate entry_ error. The problem I have is that if the **OS** already exists on the db **sqlalchemy** will try to insert it and thus generating a _duplicate entry error_. I don't understand why this is happening.
This is how I am trying to do it:

```python
def update(self, device_id: int, new_host: Host):
    with self.session_factory.begin() as session:
        device: Device = session.query(Device).options(joinedload(Device.host)).filter_by(id=device_id).one()

        db_os = session.query(OperatingSystem) \
            .filter_by(name=new_host.os.name,
                        version=new_host.os.version,
                        patch=new_host.os.patch).one_or_none()

        # If OS is in DB, add it to the new host
        if db_os:
            new_host.os = db_os

        device.host = new_host
        return True
```

An example error:

```
sqlalchemy.exc.IntegrityError: (pymysql.err.IntegrityError) (1062, "Duplicate entry 'MACOS-11.4.2-' for key 'os_uc'")
[SQL: INSERT INTO operating_system (name, version, patch) VALUES (%(name)s, %(version)s, %(patch)s)]
[parameters: {'name': 'MACOS', 'version': '11.4.2', 'patch': ''}]
```

**Comment by CaselIT** [comment]
Hi,

I've tried your script and got the same error, but I think your issue is that `db_os ` is always None. Please ensure that it's correctly found.

Note that when creating an hoes like this:
```py
Host(hostname="xx", os=OperatingSystem(name="MACOS", version="123"))
```
`OperatingSystem.path` will be `None`, not `''`, so it will not match '' on the db (unless you are on oracle)

I was able to fix using this instead
```py
Host(hostname="xx", os=OperatingSystem(name="MACOS", version="123", patch=''))
```

This is because default is not set by the init, but only after saving to the db

**Reply by santalvarez** [reply]
Hi, thanks for reaching out so quickly.
I can asure you **db_os** is not None so it is finding it on the db. Also when I create the **OperatingSystem** I explicitly pass the **patch** kwarg just like your second snippet. This error is so strange I've tried a thousand things but can't fix it.

**Reply by CaselIT** [reply]
Then I'm afraid I can't reproduce. Can you update this script to reproduce the error?

```py
from datetime import datetime
from sqlalchemy.engine.create import create_engine

from sqlalchemy.orm import (
    declarative_base,
    relationship,
    joinedload,
    sessionmaker,
)
from sqlalchemy import (
    Column,
    ForeignKey,
    UniqueConstraint,
    BigInteger,
    DateTime,
    String,
)

Base = declarative_base()

class Device(Base):
    __tablename__ = "device"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    serial_number = Column(String(100), nullable=False, unique=True)
    first_seen = Column(DateTime, nullable=False, default=datetime.utcnow)

    # One-to-one relationship with Host
    host = relationship(
        "Host",
        back_populates="device",
        uselist=False,
        cascade="all, delete-orphan",
    )

class Host(Base):
    __tablename__ = "host"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    device_id = Column(BigInteger, ForeignKey("device.id"), unique=True, nullable=False)
    os_id = Column(BigInteger, ForeignKey("operating_system.id"))
    hostname = Column(String(20), nullable=True)

    # Many Hosts to one OS
    os = relationship("OperatingSystem", back_populates="hosts")
    device = relationship("Device", back_populates="host", uselist=False)


class OperatingSystem(Base):
    __tablename__ = "operating_system"
    __table_args__ = (UniqueConstraint("name", "version", "patch", name="os_uc"),)

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    name = Column(String(49), nullable=False)
    version = Column(String(30), nullable=False)
    patch = Column(String(30), default="", nullable=False)

    # One OS to many Hosts
    hosts = relationship("Host", back_populates="os")

# change to your connection string, pg it's just an example
e = create_engine("postgresql://scott:tiger@/test", echo=True)

def update(device_id: int, new_host: Host):
    with session_factory.begin() as session:
        device: Device = (
            session.query(Device).options(joinedload(Device.host)).filter_by(id=device_id).one()
        )

        db_os = (
            session.query(OperatingSystem)
            .filter_by(
                name=new_host.os.name,
                version=new_host.os.version,
                patch=new_host.os.patch,
            )
            .one_or_none()
        )

        # If OS is in DB, add it to the new host
        if db_os:
            print("x")
            new_host.os = db_os

        device.host = new_host
        return True

Base.metadata.create_all(e)
session_factory = sessionmaker(e)
s = session_factory()
s.add(Device(id=1, serial_number="abc"))
s.add(OperatingSystem(name="foo", version="123"))
s.commit()

os = OperatingSystem(name="foo", version="123", patch="")
try:
    update(1, Host(hostname="xx", os=os))
finally:
    Base.metadata.drop_all(e)
```

**Reply by zzzeek** [reply]
> Then I'm afraid I can't reproduce. Can you update this script to reproduce the error?
> 
> ```python
> from datetime import datetime
> from sqlalchemy.engine.create import create_engine
> 
> from sqlalchemy.orm import (
>     declarative_base,
>     relationship,
>     joinedload,
>     sessionmaker,
> )
> from sqlalchemy import (
>     Column,
>     ForeignKey,
>     UniqueConstraint,
>     BigInteger,
>     DateTime,
>     String,
> )
> 
> Base = declarative_base()
> 
> class Device(Base):
>     __tablename__ = "device"
>     id = Column(BigInteger, primary_key=True, autoincrement=True)
>     serial_number = Column(String(100), nullable=False, unique=True)
>     first_seen = Column(DateTime, nullable=False, default=datetime.utcnow)
> 
>     # One-to-one relationship with Host
>     host = relationship(
>         "Host",
>         back_populates="device",
>         uselist=False,
>         cascade="all, delete-orphan",
>     )
> 
> class Host(Base):
>     __tablename__ = "host"
>     id = Column(BigInteger, primary_key=True, autoincrement=True)
>     device_id = Column(BigInteger, ForeignKey("device.id"), unique=True, nullable=False)
>     os_id = Column(BigInteger, ForeignKey("operating_system.id"))
>     hostname = Column(String(20), nullable=True)
> 
>     # Many Hosts to one OS
>     os = relationship("OperatingSystem", back_populates="hosts")
>     device = relationship("Device", back_populates="host", uselist=False)
> 
> 
> class OperatingSystem(Base):
>     __tablename__ = "operating_system"
>     __table_args__ = (UniqueConstraint("name", "version", "patch", name="os_uc"),)
> 
>     id = Column(BigInteger, primary_key=True, autoincrement=True)
>     name = Column(String(49), nullable=False)
>     version = Column(String(30), nullable=False)
>     patch = Column(String(30), default="", nullable=False)
> 
>     # One OS to many Hosts
>     hosts = relationship("Host", back_populates="os")
> 
> # change to your connection string, pg it's just an example
> e = create_engine("postgresql://scott:tiger@/test", echo=True)
> 
> def update(device_id: int, new_host: Host):
>     with session_factory.begin() as session:
>         device: Device = (
>             session.query(Device).options(joinedload(Device.host)).filter_by(id=device_id).one()
>         )
> 
>         db_os = (
>             session.query(OperatingSystem)
>             .filter_by(
>                 name=new_host.os.name,
>                 version=new_host.os.version,
>                 patch=new_host.os.patch,
>             )
>             .one_or_none()
>         )
> 
>         # If OS is in DB, add it to the new host
>         if db_os:
>             print("x")
>             new_host.os = db_os
> 
>         device.host = new_host
>         return True
> 
> Base.metadata.create_all(e)
> session_factory = sessionmaker(e)
> s = session_factory()
> s.add(Device(id=1, serial_number="abc"))
> s.add(OperatingSystem(name="foo", version="123"))
> s.commit()
> 
> os = OperatingSystem(name="foo", version="123", patch="")
> try:
>     update(1, Host(hostname="xx", os=os))
> finally:
>     Base.metadata.drop_all(e)
> ```

it reproduces in 1.4, they have a backref cascade happening

**Reply by fabfernandez** [reply]
Yes, we are using slq alchemy 1.4.46

**Reply by CaselIT** [reply]
I assumed v2 :) that's fine there

**Comment by zzzeek** [comment, marked as answer]
change your echo=True to echo='debug', then get the SQL output for your program and walk through every statement, the parameters being sent, and the results received for each.   The database isn't lying about that error, so either a. there must be an INSERT preceding it which is putting in that key or b. your test for "key already exists" is using criteria that does not match the ultimate values that get put into the database.

**Reply by fabfernandez** [reply]
Hi @zzzeek , I work with @santalvarez on this project.
Yes, we activated the logs, and we can confirm that slq-alchemy is executing an INSERT on the `os` table.
We need to stop that from happening when the os has been found.
Here are the logs:
```sql

2023-02-07 12:05:15,047 INFO sqlalchemy.engine.Engine SELECT operating_system.id AS operating_system_id, operating_system.name AS operating_system_name, operating_system.version AS operating_system_version, operating_system.patch AS operating_system_patch 
FROM operating_system 
WHERE operating_system.name = %(name_1)s AND operating_system.version = %(version_1)s AND operating_system.patch = %(patch_1)s 
 LIMIT %(param_1)s
2023-02-07 12:05:15,047 INFO sqlalchemy.engine.Engine [generated in 0.00031s] {'name_1': 'MACOS', 'version_1': '11.4.2', 'patch_1': '', 'param_1': 1}
2023-02-07 12:05:15,225 INFO sqlalchemy.engine.Engine INSERT INTO operating_system (name, version, patch) VALUES (%(name)s, %(version)s, %(patch)s)
2023-02-07 12:05:15,225 INFO sqlalchemy.engine.Engine [generated in 0.00036s] {'name': 'MACOS', 'version': '11.4.2', 'patch': ''}
2023-02-07 12:05:15,567 INFO sqlalchemy.engine.Engine ROLLBACK
```

**Reply by zzzeek** [reply]
OK, the issue is that your hand-created OperatingSystem is also being added to the Session and is not being removed.    you are using an older version of SQLAlchemy and the behavior is known as a "backref cascade", a [behavior that's removed in SQLAlchemy 2.0](https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#change-5150).

Here is how you debug an issue like this.   Look at the object you *dont* want in the Session, which is the OperatingSystem you are creating by hand.  then assert that it's not there:

```py
        old_os = new_host.os

        # If OS is in DB, add it to the new host
        if db_os:
            print("x")
            new_host.os = db_os

        device.host = new_host

        assert old_os not in session

```

running @CaselIT 's script at https://github.com/sqlalchemy/sqlalchemy/discussions/9252#discussioncomment-4887147 on SQLAlchemy 1.4, this assertion fails.


**Immediate solution number one**:

without having to figure out anything at all why the above is happening, you can fix your program immediately by removing that object:

```py
        if old_os in session:
            session.expunge(old_os)
```

the condition occurs because here:

```py
        if db_os:
            print("x")
            new_host.os = db_os
        assert not old_os in session
```

step 1: new_host.os is assigned to db_os
step 2: backref happens first, new_host is appended to db_os.hosts
step 3: backref cascade happens (the thing that is removed in 2.0) - hosts gets added to the Session
step 4: all of host.os gets added to the session.  the **old** OperatingSystem
step 5: undesired object is in the session.
step 6: old os is replaced on host.os with db_os

**Immediate solution number two**:

You can **disable the now removed behavior in older SQLAlchemy versions** by disabling backref cascade:

    # One OS to many Hosts
    hosts = relationship("Host", back_populates="os", cascade_backrefs=False)

**Longer term thing to look out for and possibly adjust practices**:

Always look for objects being cascaded into sessions.    creating a `Host(os = OperatingSystem())` strongly implies that both of those objects are going to be persisted.    For your function, I would prefer to write it as:

```py
def update(device_id: int, new_host: Host, potential_os: OperatingSystem):
    with session_factory.begin() as session:
        device: Device = (
            session.query(Device)
            .options(joinedload(Device.host))
            .filter_by(id=device_id)
            .one()
        )

        db_os = (
            session.query(OperatingSystem)
            .filter_by(
                name=potential_os.name,
                version=potential_os.version,
                patch=potential_os.patch,
            )
            .one_or_none()
        )


        if db_os:
            # If OS is in DB, add it to the new host
            new_host.os = db_os
        else:
            # otherwise use old one
            new_host.os = potential_os

        device.host = new_host

        return True

os = OperatingSystem(name="foo", version="123", patch="")
update(1, Host(hostname="xx"), os)
```

the above approach makes the intent clearer that the `OperatingSystem` created on the outside is intended for **conditional insertion only**.

**Reply by fabfernandez** [reply]
@zzzeek Mike thank you so much for your time, that worked!!!

I knew it had to do with the backref cascade, I'm gonna seriously consider using version 2.0 of alchemy

I tried adding the `cascade_backrefs=False`, but failed as I added it to the Host side of the relationship.
I still don't understand how to clearly see the correct side of the relationship to which I have to add the `cascade_backrefs=False`

To me, his behavior is difficult to understand, and I see why it's disabled on version 2.0

Again, thank you very much
Also thanks to @CaselIT for your time and effort as well.

**Reply by fabfernandez** [reply]
I'll let @santalvarez know so he can mark this as the solution asap.
